### PR TITLE
[DISCO-2427] Metrics ADR and ADR Template

### PR DESCRIPTION
**Pull request by taddes**
## References

JIRA: [DISCO-2427](https://mozilla-hub.atlassian.net/browse/DISCO-2427)
GitHub: [#136 ](https://github.com/mozilla-services/consvc-shepherd/issues/136)

## Description
Create an ADR to capture the benefits and decision behind using Markus as our metrics client library, as discussed by the team and sourced from other Mozilla teams. Can be used for future reference given the possibility of using the library on other projects.

Thanks for the suggestion to do this, @hackebrot !

## PR Review Checklist

_Put an `x` in the boxes that apply_

- [x] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/consvc-shepherd/blob/main/CONTRIBUTING.md).
- [x] The PR title starts with the JIRA issue reference, format `[DISCO-####]`, and has the same title (if applicable).
- [x] [Documentation](https://github.com/mozilla-services/consvc-shepherd/tree/main/docs) has been updated (if applicable).
- [ ] Functional and performance test coverage has been expanded and maintained (if applicable).

[DISCO-2427]: https://mozilla-hub.atlassian.net/browse/DISCO-2427?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+## Decision Drivers
+
+1. Easy to implement, provides necessary features. 
+2. Does not require excessive customization to Shepherd that hinders current/future development.
+3. Has precedent of effective usage, is used at Mozilla by other teams with more mature Django applications.
+4. Good documentation, engagement, and maintenance with broader community.
+5. Easy testing (plus if testing/mocking framework provided). 
+
+## Considered Options
```
Comment (hackebrot) [inline docs/adr/0000-metrics-client-for-shepherd.md:19]: Links to the PyPI sites for these packages would be helpful.

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+* D. aiodogstatsd
+
+## Decision Outcome
+
+Chosen option:
+
+* A. Markus
+
+Markus essentially meets all our required needs with a number of significant features that provide great benefit. 
+Most StatsD libraries are simple, in that the core functions of calling an increment, gauge, timer and histogram.  Markus matches this simplicity with added flexibility, in that you can define tags and filters when calling a given metric.
```
Comment (hackebrot) [inline docs/adr/0000-metrics-client-for-shepherd.md:33]: See above comment about using the word "simple".

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    def filter(self, record):
+        if "debug:true" in record.tags:
+            return
+        return record
+```
+
+### Positive Consequences
+
+* Quality of metrics customizations mean significant control of what metrics we want to emit.
+* Built-in DogStatsD though backend.
```
Comment (hackebrot) [inline docs/adr/0000-metrics-client-for-shepherd.md:85]:
```suggestion
* Built-in DogStatsD through backend.
```

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+* Status: accepted
+* Deciders: taddes, tif, nan
+* Date: 2023-05-30
+
+## Context and Problem Statement
+
+Shepherd is currently a simple Django Admin application used to create and publish snapshots and allocations of advertising partners.  The recent expansion of the service through the Project Honeycomb: Share-of-Voice epic [DISCO-2290] along with moving to a continuous deployment model [DISCO-2208] necessitated discussion around metrics and observability.  Up to this point, we did not have any information to analyze Shepherd's internal state pre or post deploy other than some automated testing and manual testing.
+
+## Decision Drivers
```
Comment (hackebrot) [inline docs/adr/0000-metrics-client-for-shepherd.md:11]: This list seems to be missing the async/sync point you made on Slack.

Another important driver with respect to a Rapid Release Model is that it's easier for engineers to contribute and ship code to production when we're we’re using the same tool (library/framework) for the same task (metrics collection) for similar projects (Python services developed and owned by this team).

**Review by hackebrot — CHANGES_REQUESTED** [review]
Hi @taddes! 👋🏻

Thank you for creating this!

I've added my comments inline.

**Comment by taddes** [comment]
Thanks for the thoughtful review and useful comments @hackebrot , I've made the recommended changes. Glad we'll have this to reference in our ongoing discussions regarding metrics.

**Review by hackebrot — CHANGES_REQUESTED** [review]
Thank you for addressing my comments, @taddes!

Added my feedback on the updated revision inline.

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+* D. aiodogstatsd
+
+## Decision Outcome
+
+Chosen option:
+
+* A. Markus
+
+Markus essentially meets all our required needs with a number of significant features that provide great benefit. 
+Most StatsD libraries are not excessively complicated, in that the core functions of calling an increment, gauge, timer and histogram.  Markus matches this need with added flexibility, in that you can define tags and filters when calling a given metric and define backends.
```
[the remark this answers is not in the document]
Comment (taddes) [reply docs/adr/0000-metrics-client-for-shepherd.md:35]: The examples below and the `Positive Consequences` section capture the benefits and exemplify the described Decision Outcome. I added an additional line describing why aiodogstatsd would be inappropriate given this exploration.

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+```python
+get_metrics("shepherd")
+markus.get_metrics(thing='', extra='', filters=None)
+```
+
+`thing` is a prefix to use for keys 
+`extra` adds bits to end of prefix
+`filters`, list of filters to apply
+
+**[Backends](https://markus.readthedocs.io/en/latest/backends.html)** - A significant feature Markus has over other libraries is the ability to define metrics backends. It comes with the `LoggingMetrics`, `LoggingRollupMetrics`, `StatsdMetrics`, `DatadogMetrics`, `CloudwatchMetrics` by default, handling numerous configuration optimizations. You can also define your own backedn.
```
Comment (Trinaa) [inline docs/adr/0000-metrics-client-for-shepherd.md:49]: backend?

```suggestion
**[Backends](https://markus.readthedocs.io/en/latest/backends.html)** - A significant feature Markus has over other libraries is the ability to define metrics backends. It comes with the `LoggingMetrics`, `LoggingRollupMetrics`, `StatsdMetrics`, `DatadogMetrics`, `CloudwatchMetrics` by default, handling numerous configuration optimizations. You can also define your own backend.
```

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        "class": "markus.backends.logging.LoggingMetrics",
+        "options": {
+            "logger_name": "markus",
+            "leader": "METRICS",
+        }
+    }
+])
+```
+
+**[Filters](https://markus.readthedocs.io/en/latest/filters.html)** - Markus lets you write filters to modify generated metrics on the fly. This allows infinite customization to drop or modify metric values before they are emitted. Th
```
Comment (hackebrot) [inline docs/adr/0000-metrics-client-for-shepherd.md:84]: Looks like something is missing here or the `Th` needs to be deleted.

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+* None that come to mind, there are other viable alternatives but Markus seems to fit our needs very well.
+* Possibly that Markus is a new and different library from say aiodogstatsd, which is used in Merino. However, that is an async library and it has several drabacks compared to Markus for our use case.
+
+
+## Pros and Cons of the Options
+
+### Markus
+[Docs](https://markus.readthedocs.io/en/latest/)
+
+[pypi](https://pypi.org/project/markus/)
```
Comment (hackebrot) [inline docs/adr/0000-metrics-client-for-shepherd.md:114]: Python Package Index (PyPI). All lowercase is the URL. Would be good to get this updated throughout the doc.

```suggestion
[PyPI](https://pypi.org/project/markus/)
```

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+[pypi](https://pypi.org/project/aiodogstatsd/)
+
+[GitHub Repo](https://github.com/Gr1N/aiodogstatsd)
+
+#### Pros
+
+* Already used in Merino.
+* Fairly self-evident implementation.
+
```
Comment (hackebrot) [inline docs/adr/0000-metrics-client-for-shepherd.md:193]: Unnecessary white-space.

File: docs/adr/0000-metrics-client-for-shepherd.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+* Date: 2023-05-30
+
+[@taddes]: https://github.com/taddes
+[@tiftran]: https://github.com/tiftran
+[@ncloudioj]: https://github.com/ncloudioj
+
+
+## Context and Problem Statement
+
+Shepherd is currently a minimal Django Admin application used to create and publish snapshots of advertiser settings of Sponsored Tiles. The recent expansion of the service through the Project Honeycomb: Share-of-Voice epic [DISCO-2290] along with moving to a developer-driven Continuous Deployment model [DISCO-2208] necessitated discussion around application metrics and observability. Up to this point, we did not have any information to analyze Shepherd's internal state pre or post deploy aside from automated testing and manual testing.
```
Comment (hackebrot) [inline docs/adr/0000-metrics-client-for-shepherd.md:14]: Missing links for `[DISCO-2290]` and `[DISCO-2208]`.

**Review by hackebrot — CHANGES_REQUESTED** [review]

File: docs/adr/0000-metrics-client-for-shepherd.md (hunk 10 above)
Comment (taddes) [reply docs/adr/0000-metrics-client-for-shepherd.md:14]: I hadn't intended to add the links originally, as the urls point to our internal Jira boards and I didn't necessarily want that in a public repo. Anyone internal in Mozilla could use the project and number to view them, but I'm happy to add them.

**Review by hackebrot — APPROVED** [review]
Appreciate the work you've done on this ADR, @taddes! 🚀

**Comment by taddes** [comment]
Thanks so much @hackebrot . Appreciate feedback and great that we can lean on this process to have well-documented technical decisions.

**Pull request merged**